pkgs/servers/code-server/default.nix

```diff
@@ -0,0 +1,139 @@
+{ stdenv, fetchFromGitHub, python, zip, makeWrapper, fetchurl, pkgconfig, git
+, rsync, makeDesktopItem, runtimeShell, writeScript, runCommand
+, nodejs, yarn, libsecret, xorg, ripgrep, electron }:
+
+with stdenv.lib;
+
+let
+  system = stdenv.hostPlatform.system;
+
+in stdenv.mkDerivation rec {
+  pname = "code-server";
+  version = "3.2.0";
+  commit = "fd36a99a4c78669970ebc4eb05768293b657716f";
+
+  src = fetchFromGitHub {
+    owner = "cdr";
+    repo = "code-server";
+    rev = version;
+    sha256 = "PX2WE4Wr4XUcJFw/FfKZ7PxueMXCOV/qQRGVqPThTqQ=";
+    fetchSubmodules = true;
+  };
+
+  yarnCache = stdenv.mkDerivation {
+    name = "${pname}-${version}-${system}-yarn-cache";
+    inherit src;
+    phases = ["unpackPhase" "buildPhase"];
+    nativeBuildInputs = [ yarn git ];
+    buildPhase = ''
+      export HOME=$PWD
+
+      # apply code-server patches as code-server has patched vscode yarn.lock
+      yarn vscode:patch
+
+      yarn config set yarn-offline-mirror $out
+      find "$PWD" -name "yarn.lock" -printf "%h\n" | \
+        xargs -I {} yarn install --cwd {} \
+          --frozen-lockfile --ignore-scripts --ignore-platform --ignore-engines --no-progress --non-interactive
+    '';
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+
+    # to get hash values use nix-build -A code-server.yarnPrefetchCache --argstr system <system>
+    outputHash = {
+      x86_64-linux = "197v9660yqhvazalyhc87ciy6q2ii8xxi5bc5y96bck2w0d2y8if";
+      aarch64-linux = "LiIvGuBismWSL2yV2DuKUWDjIzuIQU/VVxtiD4xJ+6Q=";
```

Comment:
I'm not sure if we want to add even more packages to `nixpkgs` utilizing this controversial pattern.

Reply:
Explain me what do you mean by that? I think this is most clean way. What is your proposed alternative?

Reply:
The issue is described here: https://github.com/NixOS/nix/issues/2270

Although I know that `yarn2nix` can be painful in some cases, is there any reason not to use it here?

Reply:
Two major issues with using generator tools are:

- large generated files (maybe ~10mb in case of code-server), which I cannot put into nixpkgs repo as they just unnecessary pollute repo.
- community maintained tool, that tries to wrap other package manager, usually in imperfect way and can present much pain for the one creating or updating the package

I think static output hashes are best tool we have. Personally I would want to throw away purity for some derivations, as I trust underlying package manager, but nix philosophy is well to be pure, but this sometimes present great pain when just trying to package something, already spending maybe days figuring out how to do it "nix way".
Sometimes I just want to go back to Debian, as sometimes I do not care for purity, as reality is imperfect and unpure.

Reply:
Also in many cases, less magic is better, less headache, more fun.

Reply:
Don't get me wrong: while I agree with the reasoning in the issue I linked above, I don't think we should prohibit this entirely, instead this should still be an option for software which is hard to package.

I mainly wanted to make sure that packagers are aware of this and using such a fixed-output drv is the best option :)

Reply:
Ok, I understand, I just don't want to make packaging too hard and confusing due to multiple layers of abstractions, for a goal that's not even that important in some cases. I had quite some fight with lang2nix tools (implementing pypi2nix long time ago and some node2nix stuff) and when I look back, it was really not worth it for 99% cases.

Regarding this package if there would be a way to store large generated files in some other storage, I could use yarn2nix, but well at the end is the same, just one is more simple. One large file with a lot of hashes or one hash that represents them all.

Reply:
What am I really missing is better tooling for static output derivations, as for example this is what I find ugly: https://github.com/NixOS/nixpkgs/pull/87258/files#diff-97ddd5942a260ac035c022c0c57de234R122

Reply:
Oh yeah, recursive nix would help here, since i could use yarn2nix as part of build process and have generation process run as part of it.

As soon as recursive nix is supported on hydra, i can rewrite it. Still we would only have one static output hash, but process of creating deterministic output would be more stable, cached and arch independent. This i think is also a solution to improve static output derivations.

Reply:
Can be recursive nix be used in this case even? I don't think the inner nix instance has access to network or can create its own derivations as far as I know.

Reply:
Regarding network I do not know, but it should be able to create derivations, since isn't that the point of recursive nix? But yes under assumption recursive nix can access network running under static output derivation, this should be possible.

Reply:
The intention of recursive nix was to abuse nix as a build system like here: https://github.com/edolstra/nix-make without paying having to setup a sandbox for each build. I assume it can access the network if the outer nix build has it, but I don't think it can have its own derivation.